Sporadic mock verification failures related to hashCode/equals on 3.12.1

Since updating from `3.11.2` to `3.12.0` (and getting similar errors on `3.12.1`), I have seen some strange sometimes-reproducible verification failures on Mockito for tests which previously passed consistently.

I can't yet establish a pattern or simple reproduction; it feels like it is somehow dependent on the order of test evaluation/usage with respect to other tests within the same JVM instance, and I have been unable to replicate locally.

There are two main classes I am seeing

1. mocks with interactions being tracked (and thus failing `verifyNoInteractions(mock)` after being put into `HashSet`s (these did not seem to previously count as an interaction, and usually do not)
    ```
    com.thoughtworks.go.server.materials.postcommit.git.GitPostCommitHookImplementerTest > shouldReturnEmptyListIfParamHasNoValueForRepoURL() FAILED
        org.mockito.exceptions.verification.NoInteractionsWanted: 
        No interactions wanted here:
        -> at com.thoughtworks.go.server.materials.postcommit.git.GitPostCommitHookImplementerTest.shouldReturnEmptyListIfParamHasNoValueForRepoURL(GitPostCommitHookImplementerTest.java:93)
        But found this interaction on mock 'gitMaterial':
        -> at java.base/java.util.HashMap.hash(HashMap.java:340)
        Actually, above is the only interaction with this mock.
            at com.thoughtworks.go.server.materials.postcommit.git.GitPostCommitHookImplementerTest.shouldReturnEmptyListIfParamHasNoValueForRepoURL(GitPostCommitHookImplementerTest.java:93)
    ```
    https://github.com/gocd/gocd/blob/b7bba59201c857efc010c3493664802f218d2bae/server/src/test-fast/java/com/thoughtworks/go/server/materials/postcommit/git/GitPostCommitHookImplementerTest.java#L81-L94

2. mocks that are not considered equal to themselves when put inside a `Collection` and then searched for 
    ```
    com.thoughtworks.go.server.materials.postcommit.git.GitPostCommitHookImplementerTest > shouldReturnListOfMaterialMatchingThePayloadURL() FAILED
        java.lang.AssertionError: 
        Expected: a collection containing <Mock for GitMaterial, hashCode: 0>
             but: was <Mock for GitMaterial, hashCode: 0>, was <Mock for GitMaterial, hashCode: 0>
            at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
            at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:8)
            at com.thoughtworks.go.server.materials.postcommit.git.GitPostCommitHookImplementerTest.shouldReturnListOfMaterialMatchingThePayloadURL(GitPostCommitHookImplementerTest.java:58)
    ```
    https://github.com/gocd/gocd/blob/b7bba59201c857efc010c3493664802f218d2bae/server/src/test-fast/java/com/thoughtworks/go/server/materials/postcommit/git/GitPostCommitHookImplementerTest.java#L41-L65

I have a handful of other examples too in different classes.

This could be confirmation bias, however since I saw some stuff was done with respect to hashCodes in #2331 I wondered if it could be related. Logging it here to see if anyone else has seen any similar weirdness or might know what is going on.

**Environment**
- Currently I _cannot_ replicate locally (MacOS Big Sur w/AdoptOpenJDK `15.0.2`) even where it is replicable on build/CI automation environment.
- The failures seem to be (somewhat) reproducible on AL2 Linux w/ Oracle JDK `15.0.1`) but I cannot distinguish a pattern, and sometimes a different combination/ordering of tests does not seem to yield the same failure on exactly the same test code.

I have the same issue after upgrade from `3.11.2` to `3.12.1`, i.e. the mock is not equal to itself.
Here is an example of the recent failure in CI https://github.com/real-logic/agrona/runs/3401033492?check_suite_focus=true:
```

BroadcastReceiverTest > shouldReceiveTwoMessagesFromBuffer() FAILED
    java.lang.AssertionError: 
    Expected: is <Mock for UnsafeBuffer, hashCode: 0>
         but: was <Mock for UnsafeBuffer, hashCode: 0>
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
        at org.agrona.concurrent.broadcast.BroadcastReceiverTest.shouldReceiveTwoMessagesFromBuffer(BroadcastReceiverTest.java:131)
```

The issue is only reproducible in CI and only on Linux but is green locally or on Windows in CI.

--- a/src/main/java/org/mockito/internal/creation/bytebuddy/TypeCachingBytecodeGenerator.java
+++ b/src/main/java/org/mockito/internal/creation/bytebuddy/TypeCachingBytecodeGenerator.java
@@ -89,11 +89,11 @@ class TypeCachingBytecodeGenerator extends ReferenceQueue<ClassLoader>
         private final Answer defaultAnswer;
 
         private MockitoMockKey(
-            Class<?> type,
-            Set<Class<?>> additionalType,
-            SerializableMode serializableMode,
-            boolean stripAnnotations,
-            Answer defaultAnswer) {
+                Class<?> type,
+                Set<Class<?>> additionalType,
+                SerializableMode serializableMode,
+                boolean stripAnnotations,
+                Answer defaultAnswer) {
             super(type, additionalType);
             this.serializableMode = serializableMode;
             this.stripAnnotations = stripAnnotations;

--- a/src/main/java/org/mockito/internal/creation/bytebuddy/TypeCachingBytecodeGenerator.java
+++ b/src/main/java/org/mockito/internal/creation/bytebuddy/TypeCachingBytecodeGenerator.java
@@ -5,12 +5,14 @@
 package org.mockito.internal.creation.bytebuddy;
 
 import java.lang.ref.ReferenceQueue;
+import java.util.Objects;
 import java.util.Set;
 import java.util.concurrent.locks.ReadWriteLock;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
 
 import net.bytebuddy.TypeCache;
 import org.mockito.mock.SerializableMode;
+import org.mockito.stubbing.Answer;
 
 class TypeCachingBytecodeGenerator extends ReferenceQueue<ClassLoader>
         implements BytecodeGenerator {
@@ -43,7 +45,8 @@ class TypeCachingBytecodeGenerator extends ReferenceQueue<ClassLoader>
                                     params.mockedType,
                                     params.interfaces,
                                     params.serializableMode,
-                                    params.stripAnnotations),
+                                    params.stripAnnotations,
+                                    params.defaultAnswer),
                             () -> bytecodeGenerator.mockClass(params),
                             BOOTSTRAP_LOCK);
         } catch (IllegalArgumentException exception) {
@@ -83,15 +86,18 @@ class TypeCachingBytecodeGenerator extends ReferenceQueue<ClassLoader>
 
         private final SerializableMode serializableMode;
         private final boolean stripAnnotations;
+        private final Answer defaultAnswer;
 
         private MockitoMockKey(
-                Class<?> type,
-                Set<Class<?>> additionalType,
-                SerializableMode serializableMode,
-                boolean stripAnnotations) {
+            Class<?> type,
+            Set<Class<?>> additionalType,
+            SerializableMode serializableMode,
+            boolean stripAnnotations,
+            Answer defaultAnswer) {
             super(type, additionalType);
             this.serializableMode = serializableMode;
             this.stripAnnotations = stripAnnotations;
+            this.defaultAnswer = defaultAnswer;
         }
 
         @Override
@@ -107,7 +113,8 @@ class TypeCachingBytecodeGenerator extends ReferenceQueue<ClassLoader>
             }
             MockitoMockKey that = (MockitoMockKey) object;
             return stripAnnotations == that.stripAnnotations
-                    && serializableMode.equals(that.serializableMode);
+                    && serializableMode.equals(that.serializableMode)
+                    && Objects.equals(defaultAnswer, that.defaultAnswer);
         }
 
         @Override
@@ -115,6 +122,7 @@ class TypeCachingBytecodeGenerator extends ReferenceQueue<ClassLoader>
             int result = super.hashCode();
             result = 31 * result + (stripAnnotations ? 1 : 0);
             result = 31 * result + serializableMode.hashCode();
+            result = 31 * result + Objects.hashCode(defaultAnswer);
             return result;
         }
     }

--- a/src/test/java/org/mockito/internal/creation/bytebuddy/TypeCachingMockBytecodeGeneratorTest.java
+++ b/src/test/java/org/mockito/internal/creation/bytebuddy/TypeCachingMockBytecodeGeneratorTest.java
@@ -14,6 +14,8 @@ import java.lang.ref.Reference;
 import java.lang.ref.ReferenceQueue;
 import java.lang.ref.WeakReference;
 import java.util.Collections;
+import java.util.IdentityHashMap;
+import java.util.Set;
 import java.util.WeakHashMap;
 
 import org.junit.Before;
@@ -140,6 +142,40 @@ public class TypeCachingMockBytecodeGeneratorTest {
                                 Answers.RETURNS_DEFAULTS));
 
         assertThat(other_mock_type).isNotSameAs(the_mock_type);
+    }
+
+    @Test
+    public void ensure_cache_returns_different_instance_defaultAnswer() throws Exception {
+        // given
+        ClassLoader classloader_with_life_shorter_than_cache =
+                inMemoryClassLoader()
+                        .withClassDefinition("foo.Bar", makeMarkerInterface("foo.Bar"))
+                        .build();
+
+        TypeCachingBytecodeGenerator cachingMockBytecodeGenerator =
+                new TypeCachingBytecodeGenerator(new SubclassBytecodeGenerator(), true);
+
+        Answers[] answers = Answers.values();
+        Set<Class<?>> classes = Collections.newSetFromMap(new IdentityHashMap<>());
+        classes.add(cachingMockBytecodeGenerator.mockClass(
+            withMockFeatures(
+                classloader_with_life_shorter_than_cache.loadClass("foo.Bar"),
+                Collections.<Class<?>>emptySet(),
+                SerializableMode.NONE,
+                false,
+                null)));
+        for (Answers answer : answers) {
+            Class<?> klass = cachingMockBytecodeGenerator.mockClass(
+                withMockFeatures(
+                    classloader_with_life_shorter_than_cache.loadClass("foo.Bar"),
+                    Collections.<Class<?>>emptySet(),
+                    SerializableMode.NONE,
+                    false,
+                    answer));
+            assertThat(classes.add(klass)).isTrue();
+        }
+
+        assertThat(classes).hasSize(answers.length + 1);
     }
 
     @Test

--- a/src/test/java/org/mockito/internal/creation/bytebuddy/TypeCachingMockBytecodeGeneratorTest.java
+++ b/src/test/java/org/mockito/internal/creation/bytebuddy/TypeCachingMockBytecodeGeneratorTest.java
@@ -157,21 +157,23 @@ public class TypeCachingMockBytecodeGeneratorTest {
 
         Answers[] answers = Answers.values();
         Set<Class<?>> classes = Collections.newSetFromMap(new IdentityHashMap<>());
-        classes.add(cachingMockBytecodeGenerator.mockClass(
-            withMockFeatures(
-                classloader_with_life_shorter_than_cache.loadClass("foo.Bar"),
-                Collections.<Class<?>>emptySet(),
-                SerializableMode.NONE,
-                false,
-                null)));
+        classes.add(
+                cachingMockBytecodeGenerator.mockClass(
+                        withMockFeatures(
+                                classloader_with_life_shorter_than_cache.loadClass("foo.Bar"),
+                                Collections.<Class<?>>emptySet(),
+                                SerializableMode.NONE,
+                                false,
+                                null)));
         for (Answers answer : answers) {
-            Class<?> klass = cachingMockBytecodeGenerator.mockClass(
-                withMockFeatures(
-                    classloader_with_life_shorter_than_cache.loadClass("foo.Bar"),
-                    Collections.<Class<?>>emptySet(),
-                    SerializableMode.NONE,
-                    false,
-                    answer));
+            Class<?> klass =
+                    cachingMockBytecodeGenerator.mockClass(
+                            withMockFeatures(
+                                    classloader_with_life_shorter_than_cache.loadClass("foo.Bar"),
+                                    Collections.<Class<?>>emptySet(),
+                                    SerializableMode.NONE,
+                                    false,
+                                    answer));
             assertThat(classes.add(klass)).isTrue();
         }
 

--- a/src/test/java/org/mockitousage/spies/MockCreationShouldNotAffectSpyTest.java
+++ b/src/test/java/org/mockitousage/spies/MockCreationShouldNotAffectSpyTest.java
@@ -1,0 +1,51 @@
+/*
+ * Copyright (c) 2007 Mockito contributors
+ * This program is made available under the terms of the MIT License.
+ */
+package org.mockitousage.spies;
+
+import org.junit.Test;
+import org.mockitoutil.TestBase;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+
+public class MockCreationShouldNotAffectSpyTest extends TestBase {
+
+    @Test
+    public void test() {
+        TestClass instance = new TestClass(42);
+
+        TestClass mock = mock(TestClass.class);
+        assertEquals(mock.hashCode(), mock.hashCode());
+        assertEquals(mock, mock);
+
+        TestClass spy = spy(instance);
+        assertEquals(instance.hashCode(), spy.hashCode());
+        assertEquals(spy, instance);
+        assertEquals(instance, spy);
+    }
+
+    static class TestClass {
+        private final long value;
+
+        TestClass(final long value)
+        {
+            this.value = value;
+        }
+
+        public boolean equals(final Object o)
+        {
+            if (!(o instanceof TestClass)) {
+                return false;
+            }
+            return value == ((TestClass)o).value;
+        }
+
+        public int hashCode()
+        {
+            return Long.hashCode(value);
+        }
+    }
+}

--- a/src/test/java/org/mockitousage/spies/MockCreationShouldNotAffectSpyTest.java
+++ b/src/test/java/org/mockitousage/spies/MockCreationShouldNotAffectSpyTest.java
@@ -30,21 +30,18 @@ public class MockCreationShouldNotAffectSpyTest extends TestBase {
     static class TestClass {
         private final long value;
 
-        TestClass(final long value)
-        {
+        TestClass(final long value) {
             this.value = value;
         }
 
-        public boolean equals(final Object o)
-        {
+        public boolean equals(final Object o) {
             if (!(o instanceof TestClass)) {
                 return false;
             }
-            return value == ((TestClass)o).value;
+            return value == ((TestClass) o).value;
         }
 
-        public int hashCode()
-        {
+        public int hashCode() {
             return Long.hashCode(value);
         }
     }

--- a/src/test/java/org/mockitousage/spies/SpyCreationShouldNotAffectMockTest.java
+++ b/src/test/java/org/mockitousage/spies/SpyCreationShouldNotAffectMockTest.java
@@ -35,11 +35,10 @@ public class SpyCreationShouldNotAffectMockTest extends TestBase {
         }
 
         public boolean equals(final Object o) {
-            if (!(o instanceof TestClass))
-            {
+            if (!(o instanceof TestClass)) {
                 return false;
             }
-            return value == ((TestClass)o).value;
+            return value == ((TestClass) o).value;
         }
 
         public int hashCode() {

--- a/src/test/java/org/mockitousage/spies/SpyCreationShouldNotAffectMockTest.java
+++ b/src/test/java/org/mockitousage/spies/SpyCreationShouldNotAffectMockTest.java
@@ -1,0 +1,49 @@
+/*
+ * Copyright (c) 2007 Mockito contributors
+ * This program is made available under the terms of the MIT License.
+ */
+package org.mockitousage.spies;
+
+import org.junit.Test;
+import org.mockitoutil.TestBase;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+
+public class SpyCreationShouldNotAffectMockTest extends TestBase {
+
+    @Test
+    public void test() {
+        TestClass instance = new TestClass(Long.MIN_VALUE);
+
+        TestClass spy = spy(instance);
+        assertEquals(instance.hashCode(), spy.hashCode());
+        assertEquals(spy, instance);
+        assertEquals(instance, spy);
+
+        TestClass mock = mock(TestClass.class);
+        assertEquals(mock.hashCode(), mock.hashCode());
+        assertEquals(mock, mock);
+    }
+
+    static class TestClass {
+        private final long value;
+
+        TestClass(final long value) {
+            this.value = value;
+        }
+
+        public boolean equals(final Object o) {
+            if (!(o instanceof TestClass))
+            {
+                return false;
+            }
+            return value == ((TestClass)o).value;
+        }
+
+        public int hashCode() {
+            return Long.hashCode(value);
+        }
+    }
+}

